File: scripts/analyze_skill_coverage.js
#!/usr/bin/env node
/**
 * Skill Coverage Analyzer (JavaScript Edition)
 * Analyzes which skills apply to which parts of the codebase.
 * 
 * Usage:
 *   node analyze_skill_coverage.js <source-directory>
 *   node analyze_skill_coverage.js --report
 * 
 * Example:
 *   node analyze_skill_coverage.js src/
 *   node analyze_skill_coverage.js --report --output coverage.json
 */

const fs = require('fs');
const path = require('path');

function getFileSkills(filePath) {
  const skills = [];

  // Language-based
  if (filePath.endsWith('.ts') || filePath.endsWith('.tsx')) {
    skills.push('typescript-standards', 'error-handling');
  } else if (filePath.endsWith('.py')) {
    skills.push('python-standards', 'error-handling');
  } else if (filePath.endsWith('.go')) {
    skills.push('go-standards', 'error-handling');
  }

  // Location-based
  if (filePath.includes('api/') || filePath.includes('routes/') || filePath.includes('endpoint')) {
    skills.push('api-design', 'security-hardening', 'auth-patterns');
  }

  if (filePath.includes('db/') || filePath.includes('models/') || filePath.includes('schema')) {
    skills.push('database-design', 'security-hardening', 'privacy-compliance');
  }

  if (filePath.includes('components/') || filePath.includes('ui/') || filePath.includes('pages/')) {
    skills.push('ui-engineering', 'accessibility-standards');
  }

  if (filePath.includes('test/') || filePath.includes('.test.') || filePath.includes('.spec.') || filePath.includes('__tests__/')) {
    skills.push('testing-strategy');
  }

  if (filePath.includes('auth/') || filePath.includes('login') || filePath.includes('session')) {
    skills.push('auth-patterns', 'security-hardening');
  }

  if (filePath.includes('lib/') || filePath.includes('utils/') || filePath.includes('helpers/')) {
    skills.push('typescript-standards', 'error-handling');
  }

  if (filePath.includes('worker/') || filePath.includes('job/') || filePath.includes('queue/')) {
    skills.push('background-jobs', 'error-handling');
  }

  // Configuration files
  if (filePath.endsWith('.yml') || filePath.endsWith('.yaml') || filePath.endsWith('.json')) {
    if (filePath.includes('.github/') || filePath.includes('ci/') || filePath.includes('cd/')) {
      skills.push('devops-pipeline', 'git-workflow');
    }
  }

  if (filePath.includes('Dockerfile') || filePath.endsWith('.dockerfile')) {
    skills.push('container-orchestration', 'security-hardening');
  }

  // Documentation
  if (filePath.endsWith('.md')) {
    skills.push('documentation-standards');
  }

  return [...new Set(skills)];
}

function analyzeCoverage(sourceDir) {
  const sourcePath = path.resolve(sourceDir);

  if (!fs.existsSync(sourcePath)) {
    return { error: `Directory ${sourceDir} not found` };
  }

  // Get all source files
  const allFiles = [];
  const extensions = ['.ts', '.tsx', '.js', '.jsx', '.py', '.go', '.rs', '.java', '.md'];

  function walkDir(dir) {
    const items = fs.readdirSync(dir);
    for (const item of items) {
      const fullPath = path.join(dir, item);
      const stat = fs.statSync(fullPath);
      const relativePath = path.relative(sourcePath, fullPath);

      if (stat.isDirectory() && !item.startsWith('.') && !item.startsWith('node_modules')) {
        walkDir(fullPath);
      } else if (stat.isFile() && extensions.some(ext => item.endsWith(ext))) {
        allFiles.push(relativePath);
      }
    }
  }

  walkDir(sourcePath);

  // Map files to skills
  const skillCoverage = {};

  for (const filePath of allFiles) {
    const applicableSkills = getFileSkills(filePath);

    for (const skill of applicableSkills) {
      if (!skillCoverage[skill]) {
        skillCoverage[skill] = {
          skill,
          filesApplicable: [],
          filesCompliant: [],
          filesViolating: [],
          coveragePercentage: 0
        };
      }
      skillCoverage[skill].filesApplicable.push(filePath);
    }
  }

  // Calculate coverage stats
  const totalFiles = allFiles.length;
  const coverageStats = {
    totalFiles,
    totalSkills: Object.keys(skillCoverage).length,
    skills: {},
    uncoveredFiles: [],
  };

  for (const [skill, info] of Object.entries(skillCoverage)) {
    coverageStats.skills[skill] = {
      applicableCount: info.filesApplicable.length,
      applicablePercentage: Math.round((info.filesApplicable.length / totalFiles) * 100 * 100) / 100,
      sampleFiles: info.filesApplicable.slice(0, 5)
    };
  }

  // Find files with no skill coverage
  const coveredFiles = new Set();
  for (const info of Object.values(skillCoverage)) {
    info.filesApplicable.forEach(f => coveredFiles.add(f));
  }

  const uncovered = allFiles.filter(f => !coveredFiles.has(f));
  coverageStats.uncoveredFiles = uncovered;
  coverageStats.uncoveredCount = uncovered.length;
  coverageStats.coveragePercentage = Math.round(((totalFiles - uncovered.length) / totalFiles) * 100 * 100) / 100;

  return coverageStats;
}

function printCoverageReport(coverage) {
  console.log('\n' + '='.repeat(70));
  console.log('  SKILL COVERAGE ANALYSIS');
  console.log('='.repeat(70));

  if (coverage.error) {
    console.log(`\n  ❌ Error: ${coverage.error}`);
    return;
  }

  console.log(`\n  Total Files Analyzed: ${coverage.totalFiles}`);
  console.log(`  Total Skills Active: ${coverage.totalSkills}`);
  console.log(`  Overall Coverage: ${coverage.coveragePercentage}%`);
  console.log(`  Uncovered Files: ${coverage.uncoveredCount}`);

  console.log('\n' + '-'.repeat(70));
  console.log('  COVERAGE BY SKILL:');
  console.log('-'.repeat(70));

  // Sort by applicable count
  const sortedSkills = Object.entries(coverage.skills)
    .sort((a, b) => b[1].applicableCount - a[1].applicableCount);

  for (const [skill, stats] of sortedSkills) {
    const barLength = Math.floor(stats.applicablePercentage / 2);
    const bar = '█'.repeat(barLength) + '░'.repeat(50 - barLength);
    console.log(`\n  ${skill.padEnd(30)} ${String(stats.applicableCount).padStart(4)} files (${String(stats.applicablePercentage).padStart(5)}%)`);
    console.log(`  ${bar}`);
    if (stats.sampleFiles.length > 0) {
      console.log(`    Examples: ${stats.sampleFiles.slice(0, 3).join(', ')}`);
    }
  }

  if (coverage.uncoveredFiles.length > 0) {
    console.log('\n' + '-'.repeat(70));
    console.log('  UNCOVERED FILES (No skills apply):');
    console.log('-'.repeat(70));
    for (const f of coverage.uncoveredFiles.slice(0, 10)) {
      console.log(`    • ${f}`);
    }
    if (coverage.uncoveredFiles.length > 10) {
      console.log(`    ... and ${coverage.uncoveredFiles.length - 10} more`);
    }
  }

  console.log('\n' + '='.repeat(70));

  // Recommendations
  console.log('\n  RECOMMENDATIONS:');
  if (coverage.uncoveredCount > 0) {
    console.log('  ⚠️  Some files have no skill coverage - consider adding relevant skills');
  }

  const lowCoverageSkills = Object.entries(coverage.skills)
    .filter(([skill, stats]) => stats.applicableCount < 5);
  
  if (lowCoverageSkills.length > 0) {
    console.log('  ⚠️  Low coverage skills:');
    for (const [skill, stats] of lowCoverageSkills) {
      console.log(`     • ${skill}: only ${stats.applicableCount} files`);
    }
  }

  console.log('='.repeat(70) + '\n');
}

function main() {
  const args = process.argv.slice(2);
  
  let sourceDir = 'src';
  let outputFile = null;
  let specificSkill = null;
  let generateReport = false;

  // Parse arguments
  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--report') {
      generateReport = true;
    } else if (args[i] === '--output' && i + 1 < args.length) {
      outputFile = args[i + 1];
      i++;
    } else if (args[i] === '--skill' && i + 1 < args.length) {
      specificSkill = args[i + 1];
      i++;
    } else if (!args[i].startsWith('--')) {
      sourceDir = args[i];
    }
  }

  console.log(`🔍 Analyzing skill coverage in ${sourceDir}...`);

  const coverage = analyzeCoverage(sourceDir);

  if (specificSkill) {
    if (coverage.skills && coverage.skills[specificSkill]) {
      const skillData = coverage.skills[specificSkill];
      console.log(`\n📊 Coverage for ${specificSkill}:`);
      console.log(`   Applicable to: ${skillData.applicableCount} files`);
      console.log(`   Percentage: ${skillData.applicablePercentage}%`);
      console.log(`\n   Sample files:`);
      for (const f of skillData.sampleFiles.slice(0, 10)) {
        console.log(`     • ${f}`);
      }
    } else {
      console.log(`\n❌ Skill '${specificSkill}' not found or has no coverage`);
    }
  } else if (outputFile) {
    fs.writeFileSync(outputFile, JSON.stringify(coverage, null, 2));
    console.log(`\n✅ Coverage report saved to ${outputFile}`);
  } else {
    printCoverageReport(coverage);
  }
}

if (require.main === module) {
  main();
}

module.exports = { analyzeCoverage, getFileSkills };

File: scripts/analyze_skill_coverage.py
#!/usr/bin/env python3
"""
Skill Coverage Analyzer
Analyzes which skills apply to which parts of the codebase.

Usage:
    python analyze_skill_coverage.py <source-directory>
    python analyze_skill_coverage.py --report

Example:
    python analyze_skill_coverage.py src/
    python analyze_skill_coverage.py --report --output coverage.json
"""

import argparse
import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional


@dataclass
class CoverageInfo:
    skill: str
    files_applicable: list[str] = field(default_factory=list)
    files_compliant: list[str] = field(default_factory=list)
    files_violating: list[str] = field(default_factory=list)
    coverage_percentage: float = 0.0


def get_file_skills(file_path: str) -> list[str]:
    """Determine which skills apply to a file."""
    skills = []

    # Language-based
    if file_path.endswith((".ts", ".tsx")):
        skills.extend(["typescript-standards", "error-handling"])
    elif file_path.endswith(".py"):
        skills.extend(["python-standards", "error-handling"])
    elif file_path.endswith(".go"):
        skills.extend(["go-standards", "error-handling"])

    # Location-based
    if any(x in file_path for x in ["api/", "routes/", "endpoint"]):
        skills.extend(["api-design", "security-hardening", "auth-patterns"])

    if any(x in file_path for x in ["db/", "models/", "schema"]):
        skills.extend(["database-design", "security-hardening", "privacy-compliance"])

    if any(x in file_path for x in ["components/", "ui/", "pages/"]):
        skills.extend(["ui-engineering", "accessibility-standards"])

    if any(x in file_path for x in ["test/", ".test.", ".spec.", "__tests__/"]):
        skills.append("testing-strategy")

    if any(x in file_path for x in ["auth/", "login", "session"]):
        skills.extend(["auth-patterns", "security-hardening"])

    if any(x in file_path for x in ["lib/", "utils/", "helpers/"]):
        skills.extend(["typescript-standards", "error-handling"])

    if any(x in file_path for x in ["worker/", "job/", "queue/"]):
        skills.extend(["background-jobs", "error-handling"])

    # Configuration files
    if file_path.endswith((".yml", ".yaml", ".json")):
        if any(x in file_path for x in [".github/", "ci/", "cd/"]):
            skills.extend(["devops-pipeline", "git-workflow"])

    if "Dockerfile" in file_path or file_path.endswith(".dockerfile"):
        skills.extend(["container-orchestration", "security-hardening"])

    # Documentation
    if file_path.endswith(".md"):
        skills.append("documentation-standards")

    return list(set(skills))


def analyze_coverage(source_dir: str) -> dict:
    """Analyze skill coverage for a codebase."""
    source_path = Path(source_dir)

    if not source_path.exists():
        return {"error": f"Directory {source_dir} not found"}

    # Get all source files
    all_files = []
    for ext in [".ts", ".tsx", ".js", ".jsx", ".py", ".go", ".rs", ".java", ".md"]:
        all_files.extend(source_path.rglob(f"*{ext}"))

    # Map files to skills
    skill_coverage = {}

    for file_path in all_files:
        file_str = str(file_path.relative_to(source_path))
        applicable_skills = get_file_skills(file_str)

        for skill in applicable_skills:
            if skill not in skill_coverage:
                skill_coverage[skill] = CoverageInfo(skill=skill)
            skill_coverage[skill].files_applicable.append(file_str)

    # Calculate coverage stats
    total_files = len(all_files)
    coverage_stats = {
        "total_files": total_files,
        "total_skills": len(skill_coverage),
        "skills": {},
        "uncovered_files": [],
    }

    for skill, info in skill_coverage.items():
        coverage_stats["skills"][skill] = {
            "applicable_count": len(info.files_applicable),
            "applicable_percentage": round(
                len(info.files_applicable) / total_files * 100, 2
            ),
            "sample_files": info.files_applicable[:5],  # First 5 examples
        }

    # Find files with no skill coverage
    covered_files = set()
    for info in skill_coverage.values():
        covered_files.update(info.files_applicable)

    all_file_strs = {str(f.relative_to(source_path)) for f in all_files}
    uncovered = all_file_strs - covered_files
    coverage_stats["uncovered_files"] = list(uncovered)
    coverage_stats["uncovered_count"] = len(uncovered)
    coverage_stats["coverage_percentage"] = round(
        (total_files - len(uncovered)) / total_files * 100, 2
    )

    return coverage_stats


def print_coverage_report(coverage: dict):
    """Print formatted coverage report."""
    print("\n" + "=" * 70)
    print("  SKILL COVERAGE ANALYSIS")
    print("=" * 70)

    if "error" in coverage:
        print(f"\n  ❌ Error: {coverage['error']}")
        return

    print(f"\n  Total Files Analyzed: {coverage['total_files']}")
    print(f"  Total Skills Active: {coverage['total_skills']}")
    print(f"  Overall Coverage: {coverage['coverage_percentage']}%")
    print(f"  Uncovered Files: {coverage['uncovered_count']}")

    print("\n" + "-" * 70)
    print("  COVERAGE BY SKILL:")
    print("-" * 70)

    # Sort by applicable count
    sorted_skills = sorted(
        coverage["skills"].items(),
        key=lambda x: x[1]["applicable_count"],
        reverse=True,
    )

    for skill, stats in sorted_skills:
        bar_length = int(stats["applicable_percentage"] / 2)
        bar = "█" * bar_length + "░" * (50 - bar_length)
        print(
            f"\n  {skill:30s} {stats['applicable_count']:4d} files ({stats['applicable_percentage']:5.1f}%)"
        )
        print(f"  {bar}")
        if stats["sample_files"]:
            print(f"    Examples: {', '.join(stats['sample_files'][:3])}")

    if coverage["uncovered_files"]:
        print("\n" + "-" * 70)
        print("  UNCOVERED FILES (No skills apply):")
        print("-" * 70)
        for f in coverage["uncovered_files"][:10]:  # Show first 10
            print(f"    • {f}")
        if len(coverage["uncovered_files"]) > 10:
            print(f"    ... and {len(coverage['uncovered_files']) - 10} more")

    print("\n" + "=" * 70)

    # Recommendations
    print("\n  RECOMMENDATIONS:")
    if coverage["uncovered_count"] > 0:
        print(
            "  ⚠️  Some files have no skill coverage - consider adding relevant skills"
        )

    low_coverage_skills = [
        (skill, stats)
        for skill, stats in coverage["skills"].items()
        if stats["applicable_count"] < 5
    ]
    if low_coverage_skills:
        print("  ⚠️  Low coverage skills:")
        for skill, stats in low_coverage_skills:
            print(f"     • {skill}: only {stats['applicable_count']} files")

    print("=" * 70 + "\n")


def main():
    parser = argparse.ArgumentParser(description="Analyze skill coverage in codebase")
    parser.add_argument(
        "source_dir",
        nargs="?",
        default="src",
        help="Source directory to analyze (default: src)",
    )
    parser.add_argument(
        "--report",
        action="store_true",
        help="Generate coverage report",
    )
    parser.add_argument(
        "--output",
        help="Save report to JSON file",
    )
    parser.add_argument(
        "--skill",
        help="Analyze coverage for specific skill only",
    )

    args = parser.parse_args()

    print(f"🔍 Analyzing skill coverage in {args.source_dir}...")

    coverage = analyze_coverage(args.source_dir)

    if args.skill:
        if args.skill in coverage.get("skills", {}):
            skill_data = coverage["skills"][args.skill]
            print(f"\n📊 Coverage for {args.skill}:")
            print(f"   Applicable to: {skill_data['applicable_count']} files")
            print(f"   Percentage: {skill_data['applicable_percentage']}%")
            print(f"\n   Sample files:")
            for f in skill_data["sample_files"][:10]:
                print(f"     • {f}")
        else:
            print(f"\n❌ Skill '{args.skill}' not found or has no coverage")
    elif args.output:
        with open(args.output, "w") as f:
            json.dump(coverage, f, indent=2)
        print(f"\n✅ Coverage report saved to {args.output}")
    else:
        print_coverage_report(coverage)


if __name__ == "__main__":
    main()
